page_cache: evict non-dirty pages

alecco opened the pull request:
If tail is dirty, keep looking backwards to evict.
While there, evict as many pages as necessary until cache is at capacity since a previous pass could have failed to evict pages.

pereman2 reviewed (commented):
Can we have some testing here? Thanks btw

alecco commented:
Changes in https://github.com/tursodatabase/limbo/commit/0511001d30d0582a317d35e910f1e48ff0b87b01:

1. fixed not evicting locked pages (bug in previous implementation)
2. moved `evict()` to beginning of `insert()` or it could evict the inserted page (bug in prev version of this PR)
3. `evict()` returns a boolean to indicate if it succeeded to make room
4. left TODO comment for failing `insert_page` when couldn't make room in cache
5. improved test helper `insert_page()` to support inserting dirty and locked pages
6. added tests for dirty and clean pages
7. add `#[must_use]` to new functions

alecco commented:
Changes in https://github.com/tursodatabase/limbo/commit/824985d3d6d511ac24000a3b9b5edde30c2bf179:

1. fixed `fmt` (sorry I'm new to rust)

alecco commented:
I can't repro locally the test failure. Is it possible to do a run of it in CI with backtrace enabled?

```
Testing: testvfs not loaded
Testing: testvfs extension loaded
thread 'main' panicked at core/storage/pager.rs:61:38:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Test FAILED: Error encountered in Limbo shell.
make: *** [Makefile:70: test-extensions] Error 1
Error: Process completed with exit code 2.
```

alecco commented:
I think there's a race with `detach()` taking the page content, but don't know why it wasn't a problem before. There are 73 calls to `Page::get_contents()`. Having a look at unsafe `Page` handling.

alecco commented:
Moved to https://github.com/tursodatabase/limbo/pull/1368
